Sming/Components/crypto/src/blake2s.cpp

```diff
@@ -0,0 +1,189 @@
+			state.h[i] ^= v[i] ^ v[i + 8];
+		}
+	}
+
+private:
+	template <size_t M0, size_t M1, size_t M2, size_t M3, size_t M4, size_t M5, size_t M6, size_t M7, size_t M8,
+			  size_t M9, size_t M10, size_t M11, size_t M12, size_t M13, size_t M14, size_t M15>
+	void shuffleRound()
+	{
+		mPerm[0] = m[M0];
+		mPerm[1] = m[M1];
+		mPerm[2] = m[M2];
+		mPerm[3] = m[M3];
+		mPerm[4] = m[M4];
+		mPerm[5] = m[M5];
+		mPerm[6] = m[M6];
+		mPerm[7] = m[M7];
+		mPerm[8] = m[M8];
+		mPerm[9] = m[M9];
+		mPerm[10] = m[M10];
+		mPerm[11] = m[M11];
+		mPerm[12] = m[M12];
+		mPerm[13] = m[M13];
+		mPerm[14] = m[M14];
+		mPerm[15] = m[M15];
+		round();
+	}
+
+	void round()
+	{
+		roundPart<0>(v[0], v[4], v[8], v[12]);
+		roundPart<1>(v[1], v[5], v[9], v[13]);
+		roundPart<2>(v[2], v[6], v[10], v[14]);
+		roundPart<3>(v[3], v[7], v[11], v[15]);
+		roundPart<4>(v[0], v[5], v[10], v[15]);
+		roundPart<5>(v[1], v[6], v[11], v[12]);
+		roundPart<6>(v[2], v[7], v[8], v[13]);
+		roundPart<7>(v[3], v[4], v[9], v[14]);
+	}
+
+	template <size_t I> void roundPart(uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d)
+	{
+		a += b + mPerm[2 * I];
+		d = Internal::ROTR(d ^ a, 16);
+		c += d;
+		b = Internal::ROTR(b ^ c, 12);
+		a += b + mPerm[2 * I + 1];
+		d = Internal::ROTR(d ^ a, 8);
+		c += d;
+		b = Internal::ROTR(b ^ c, 7);
+	}
+
+	uint32_t mPerm[16];
+	uint32_t v[16];
+	uint32_t m[16];
+	State& state;
+};
+
+} // namespace
+
+void Blake2sEngine::initCommon(size_t hashSize, size_t keySize)
+{
+	assert(hashSize > 0 && hashSize <= hashsize);
+	assert(keySize <= maxkeysize);
+
+	state = {};
+	state.hashSize = hashSize;
+
+	std::copy_n(initVectors, 8, state.h);
+	state.h[0] ^= (0x01010000U | (keySize << 8) | hashSize);
+}
+
+void Blake2sEngine::init(const Secret& key, size_t hashSize)
+{
+	initCommon(hashSize, key.size());
+
+	uint8_t block[blocksize] = {0};
+	memcpy(block, key.data(), key.size());
+	update(block, blocksize);
+	Internal::clean(block);
+}
+
+void Blake2sEngine::update(const void* data, size_t size)
+{
+	if(size == 0) {
+		return;
+	}
+
+	const uint8_t* pData = reinterpret_cast<const uint8_t*>(data);
+	Compressor<State> compressor(state);
+
+	const size_t missing = blocksize - state.bufferLength;
+	if(size > missing) {
+		memcpy(state.buffer + state.bufferLength, pData, missing);
+		compressor(state.buffer, blocksize);
+		state.bufferLength = 0;
+		pData += missing;
+		size -= missing;
+	}
+
+	if(size > blocksize) {
+		size_t numBlocks = size / blocksize;
+		if(numBlocks * blocksize == size) {
+			--numBlocks; // exclude last block if size is a multiple of the block size
+		}
+		size -= blocksize * numBlocks;
+		for(; numBlocks > 0; --numBlocks) {
+			compressor(pData, blocksize);
+			pData += blocksize;
+		}
+	}
+
+	if(size > 0) {
+		assert(state.bufferLength + size <= blocksize);
+		memcpy(state.buffer + state.bufferLength, pData, size);
+		state.bufferLength += size;
+	}
+}
+
+void Blake2sEngine::final(uint8_t* hash)
+{
+	state.f[0] = 0xFFFFFFFFU;
```

Comment:
state.f[1] ?

Reply:
`f[1]` is initialized to zero by `state = {}`. 
However, it is never changed and only used in one xor operation where it has no effect. => Can be removed completely.